### ISSUE TITLE
Search fails in mate 3 position - outputs -128.00 score but gives wrong line

### ISSUE THREAD

**Issue by punter11235** [LeelaChessZero/lc0#910]
Consider this position:
1kr2b2/1p3p2/p5pp/r7/3Nn1PN/3BPP1P/bP6/K2R2R1 b - - 4 29

It takes Lc0 with the newest T40 network (same with JHortos 320x24b although that is a tad faster) around 5 seconds  on 2080ti with cudnn-16 backend to show -128.00 score but the line it shows is:
1.  4 [-128.00]  29.... Rc1+ 30.Rxc1 Bc4+ 31.Kb1 Bxd3+ 32.Rc2 Nc5 33.Rd1 Ra4 

Which according to Leela itself is far from winning (-1.5 score).
29...Rc1+ 30.Rxc1 Bb3+ 31.Kb1 Nd2X is a simple mate in 3.
There are two issues here:
1)It takes a long time for it to find the mate. The policy values for the moves are 0.6% for the first one, 12.5% for 2nd one and 33.76% for the mate; all white's moves are the only moves so with the exception of the first move those are main policy choices. It really shouldn't take 5 seconds on 2080ti to find this very simple line.

2)Even when it already realizes the position is a forced win (-128.00 score) it still shows wrong variation which doesn't lead to a win for black.

While I am sure 1) is very undesirable and suggests a bug in search and not general MCTS/network problems 2) is just wrong output so a clear bug.

**Comment by fohristiwhirl** [comment]
Agree that it's showing a weird main line, reproduced with 42747:

```
HALTED Nodes: 574k, N/s: 34.0k, Time: 16s
100.0% Rc1+ Rxc1 Bc4+ Kb1 Bxd3+ Rc2 Nd2+ Kc1 Ra1+ Kxd2 Rxg1 (N: 359k, P: 0.24%)
```

**Comment by punter11235** [comment]
Consider also this position from a famous Polugaevsky vs. Nezhmetdinov game:
r5k1/ppp4p/3p3b/3Pn3/1n1KP1p1/1P3rP1/PBB1N2R/R2Q4 b - - 2 26

Leela reports:
1. 30 [-2.15]  26.... c6 27.dxc6 b5 28.Bd3 Nexc6+ 29.Kc3 Bg7+ 30.Kd2 Rxd3+ 31.Ke1 Rxd1+ 32.Rxd1 Bxb2 33.Rxd6 Ne5 34.Kd1 Nxa2 35.Rh5 Nc3+ 36.Nxc3 Bxc3 37.Rg5+ Kf7 38.Rh5 Kg7 39.Rg5+ Kh8 40.Rd5 Re8 41.Rxb5 h6 42.Rf5 Re7 43.Rc5 Ba1 44.Ra5 Bd4 45.Rd5 Bc3 46.Rc5 Bb2 47.Kc2 Ba1 48.Ra5 Bd4 49.Rd5 Nf3 50.Rxf3 gxf3 51.Rxd4 f2 52.Rd1 Rxe4 53.Rf1 Re2+ 54.Kd3 Rb2 55.Kc3 
2. 30 [-0.46]  26.... c5+ 27.dxc6 bxc6 28.Bd3 Bg7 29.Nf4 Nbxd3 30.Kc3 Nf2+ 31.Kc2 Nxd1 32.Rxd1 Rxg3 33.Rxd6 Nf3 34.Rh5 Ne1+ 35.Kb1 Bxb2 36.Kxb2 Rf8 37.Rg5+ Kh8 38.Rd7 

While c6 and c5 lead to the same position after forced 27.dxc6.
Is this necessary evil of having MCTS or can it be fixed somehow?

**Comment by koedem** [comment]
@punter11235 the second one is a known behavior, we don't have a transposition table or similar. (though there are ongoing attempts to change that)

The first one I'm pretty sure is because a node is made terminal if a child node is a win for us, however that child node might not have the most visits thus not make it into the PV. I've mentioned that in the discord before but people seemed to not really bothered by it.

**Comment by JMA72** [comment]
Another example for a failed mate in 3 and a wrong main line:
![image](https://user-images.githubusercontent.com/46564651/70221342-72d70600-1748-11ea-9a07-6b1c18124fad.png)
The mate (1. Rg8+ Kd7 2.Nf7 any move 3.Rd8 mate) isn't found after half an hour with a RTX 2060. And LC0 prefers material gain in the main line above a mate in 4 in the second line.

**Comment by Mardak** [comment]
The showing wrong PV of not actual mating line was fixed with 28d302f.

Before 28d302f:
```
position fen  1kr2b2/1p3p2/p5pp/r7/3Nn1PN/3BPP1P/bP6/K2R2R1 b - - 4 29
go nodes 300 searchmoves c8c1

… nodes 300 score cp 12800 pv c8c1 d1c1 a2c4 a1b1 c4d3 c1c2 e4d2 b1c1 d2f1
```

After 28d302f (included in latest release v0.23.1):
```
… nodes 300 score cp 12800 pv c8c1 d1c1 a2b3 a1b1 e4d2
```

**Comment by Mardak** [comment]
And similarly for the position from https://github.com/LeelaChessZero/lc0/issues/910#issuecomment-562042905

Before:
```
position fen r1bk3N/ppp3Rp/2np4/1B6/3P4/8/PP3P1P/4RK2 w - - 0 1
go nodes 5000 searchmoves g7g8

… nodes 5000 score cp 12800 pv g7g8 d8d7 h8f7 h7h6 h2h3 b7b6 h3h4 c8b7 g8a8
```

After:
```
… nodes 5000 score cp 12800 pv g7g8 d8d7 h8f7 h7h6 g8d8
```
